geoh5py/objects/surveys/magnetotellurics.py

```diff
@@ -0,0 +1,229 @@
+
+            if metadata is None:
+                self.default_metadata["EM Dataset"]["Receivers"] = str(self.uid)
+                metadata = self.default_metadata
+
+            self._metadata = metadata
+        return self._metadata
+
+    @metadata.setter
+    def metadata(self, values: dict):
+
+        if not isinstance(values, dict):
+            raise TypeError("'metadata' must be of type 'dict'")
+
+        if "EM Dataset" not in values:
+            raise KeyError("'EM Dataset' must be a 'metadata' key")
+
+        for key in self.default_metadata["EM Dataset"]:
+            if key not in values["EM Dataset"]:
+                raise KeyError(f"{key} argument missing from the input metadata.")
+
+        self._metadata = values
+        self.modified_attributes = "metadata"
+
+    @property
+    def survey_type(self):
+        """Type of EM survey"""
+        if getattr(self, "_survey_type", None) is None:
+            self._survey_type = self.metadata["EM Dataset"]["Survey type"]
+
+        return self._survey_type
+
+    @property
+    def unit(self):
+        """Data unit"""
+        if getattr(self, "_unit", None) is None:
+            self._unit = self.metadata["EM Dataset"]["Unit"]
+
+        return self._unit
+
+    def add_frequency_data(self, data: dict) -> Data | list[Data]:
+        """
+        Adapted from :func:`~geoh5py.objects.object_base.ObjectBase.add_data` method.
+
+        Add data per component at every frequency defined in
+        :attr:`~geoh5py.objects.surveys.magnetotellurics.Magnetotellurics.channels`.
+        Data properties such as 'values' and 'entity_type' must be provided as a
+        dictionary under each frequency such as:
+
+        .. code-block:: python
+
+            data = {
+                "Zxx (real)": {
+                    freq_1: {'values': [v_11, v_12, ...]},
+                    freq_2: {'values': [v_21, v_22, ...]},
+                    },
+                },
+                "Zxx (imaginary)": {
+                    freq_1: {
+                        'values': [v_11, v_12, ...],
+                        "entity_type": entity_type_A,
+                        ...,
+                    },
+                    freq_2: {...},
+                },
+            }
+
+        Data values association is always assumed to be 'VERTEX' and name set
+        by the component and frequency value.
+        A :obj:`geoh5py.groups.property_group.PropertyGroup` for the component gets created
+        by default to group all frequencies.
+
+        :param data: Dictionary of data to be added to the object
+
+        :return: List of new Data objects.
+        """
+        data_objects = []
+        if self.channels is None or not self.channels:
+            raise AttributeError(
+                "The 'channels' property defining frequencies must be set before adding data."
+            )
+
+        if not isinstance(data, dict):
+            raise TypeError(
```

Comment:
Would be nice to put a more thorough validator here to check the component names, nesting structure, etc.  Maybe once we get the validator module moved over?

Reply:
Oh, I see, you're doing the extra level validation in the for loop..  Great!